### PR TITLE
Add cosmosis and cosmosis-build-standard-library recipes

### PR DESCRIPTION
**Pull request by joezuntz**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

File: recipes/cosmosis-build-standard-library/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  host:
+    - python
+    - pip
+  run:
+    - python
+    - cosmosis
+    # For compiling standard librarymodules
+    - {{ compiler('c') }}
+    - {{ compiler('fortran') }}
+    - {{ compiler('cxx') }}
```
Comment (beckermr) [inline recipes/cosmosis-build-standard-library/meta.yaml:30]: I'd probably depend directly on `compilers` here.

File: recipes/cosmosis-build-standard-library/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+build:
+  number: 0
+  skip: True # [(py < 36) or win]
+  script: "{{ PYTHON }} -m pip install . -vv"
+
+requirements:
+  build:
+    - python
+    - pip
```
Comment (beckermr) [inline recipes/cosmosis-build-standard-library/meta.yaml:20]: No need for this section for this python recipe.

File: recipes/cosmosis-build-standard-library/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+
+test:
+  commands:
+    - cosmosis-build-standard-library && cd cosmosis-standard-library && cosmosis demos/demo1.ini && cosmosis-postprocess demos/demo1.ini
+    - test -f matter_power.png
+
+about:
+  home: https://github.com/joezuntz/cosmosis
+  license: BSD-2-Clause
```
Comment (beckermr) [inline recipes/cosmosis-build-standard-library/meta.yaml:53]: If the pypi package vendors code from other authors, even if that code is not compiled, you will need to put the licenses here. Even if the pypi package has the licenses, we need them in a place conda will grab them and package them. If you store them in some spot upon install, that will work. Files that are simply in the pypi sdist may or may not make it.

File: recipes/cosmosis/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+requirements:
+  build:
+    - python
+    - pip
+    - {{ compiler('c') }}
+    - {{ compiler('fortran') }}
+    - {{ compiler('cxx') }}
+    - liblapack
+    - libblas
+    - {{ mpi }}
```
Comment (beckermr) [inline recipes/cosmosis/meta.yaml:26]: This section should only have the compilers IIRC.

File: recipes/cosmosis/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    - test -f ${SP_DIR}/cosmosis/samplers/multinest/multinest_src/libnest3_mpi.so
+    - test -f ${SP_DIR}/cosmosis/samplers/polychord/polychord_src/libchord.so
+    - test -f ${SP_DIR}/cosmosis/samplers/polychord/polychord_src/libchord_mpi.so
+    - source cosmosis-configure && pytest --pyargs cosmosis
+  requires:
+    - pytest
+
+about:
+  home: https://github.com/joezuntz/cosmosis
+  license: BSD-2-Clause
```
Comment (beckermr) [inline recipes/cosmosis/meta.yaml:70]: Same comment on licenses applies here too.

**Review by beckermr — CHANGES_REQUESTED** [review]
Hi Joe! Some initial comments. Looks good so far!

File: recipes/cosmosis-build-standard-library/meta.yaml (hunk 1 above)
Comment (joezuntz) [reply recipes/cosmosis-build-standard-library/meta.yaml:27]: I tried this today but got a large number of "Overlinking" errors for all the libs that cosmosis supplied, so I've left as-is.  I managed to forget to copy them before closing the terminal - let me know if you'd like me to run again and get them!

File: recipes/cosmosis-build-standard-library/meta.yaml (hunk 3 above)
Comment (joezuntz) [reply recipes/cosmosis-build-standard-library/meta.yaml:50]: cosmosis-build-standard-library doesn't include any vendored code yet (or almost any code at all really), so nothing to do here. But I've done that below for the other package.

Licenses might be a significant blocker for packaging the CSL.  The Planck lib (among others) has no license included, just an "All rights reserved" copyright statement in the files.

File: recipes/cosmosis/meta.yaml (hunk 4 above)
Comment (joezuntz) [reply recipes/cosmosis/meta.yaml:26]: Same problem as above - I have left as-is for now.

File: recipes/cosmosis-build-standard-library/meta.yaml (hunk 1 above)
Comment (beckermr) [reply recipes/cosmosis-build-standard-library/meta.yaml:27]: I'm confused by this. Can you submit the change and we can see what happens in the CI?

File: recipes/cosmosis/meta.yaml (hunk 4 above)
Comment (beckermr) [reply recipes/cosmosis/meta.yaml]: Something is wrong then. Let's make the change and see what is going on.

File: recipes/cosmosis-build-standard-library/meta.yaml (hunk 1 above)
Comment (joezuntz) [reply recipes/cosmosis-build-standard-library/meta.yaml]: Done

File: recipes/cosmosis/meta.yaml [hunk 6]
```diff
@@ -18,9 +18,7 @@ requirements:
   build:
     - python
     - pip
-    - {{ compiler('c') }}
-    - {{ compiler('fortran') }}
-    - {{ compiler('cxx') }}
+    - compilers
     - liblapack
     - libblas
     - {{ mpi }}
```
Comment (beckermr) [inline recipes/cosmosis/meta.yaml]: Ack Sorry I was not clear. This section should only be

```
    - {{ compiler('c') }}
    - {{ compiler('fortran') }}
    - {{ compiler('cxx') }}
```

Comment (beckermr) [reply recipes/cosmosis/meta.yaml]: The `compilers` package is for use in runtime environments and adds extra symlinks to standard names for the compilers which people expect.

File: recipes/cosmosis-build-standard-library/meta.yaml (hunk 3 above)
Comment (beckermr) [reply recipes/cosmosis-build-standard-library/meta.yaml:48]: Yep we can't ship unlicensed code.

File: recipes/cosmosis/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+  script: "{{ PYTHON }} -m pip install . -vv"
+
+requirements:
+  build:
+    - python
+    - pip
+    - compilers
+    - liblapack
+    - libblas
+    - {{ mpi }}
```
Comment (beckermr) [inline recipes/cosmosis/meta.yaml:24]:
```suggestion
    - {{ compiler('c') }}
    - {{ compiler('fortran') }}
    - {{ compiler('cxx') }}```

File: recipes/cosmosis/meta.yaml (hunk 6 above)
Comment (beckermr) [reply recipes/cosmosis/meta.yaml]: I went ahead and made this change.

File: recipes/cosmosis/meta.yaml (hunk 7 above)
Comment (joezuntz) [reply recipes/cosmosis/meta.yaml]: Thanks for fixing!  The MPI compiler is needed for polychord / multinest - should that not be in there?

Comment (joezuntz) [reply recipes/cosmosis/meta.yaml]: (Needed for full functionality; it will compile without)

Comment (beckermr) [reply recipes/cosmosis/meta.yaml]: Ah the mpi compiler. I don't think just having {{ mpi }} in there will get that compiler. You'll want things like mpich-mpicc etc.

Comment (joezuntz) [reply recipes/cosmosis/meta.yaml]: It seemed to include them when I ran this locally. Or something included them.  You think there should be another package as well?

File: recipes/cosmosis/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+build:
+  number: 0
+  skip: True  # [(py < 36) or win]
+  script: "{{ PYTHON }} -m pip install . -vv"
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+    - {{ compiler('fortran') }}
+    - {{ compiler('cxx') }}```
```
Comment (beckermr) [inline recipes/cosmosis/meta.yaml:21]:
```suggestion
    - {{ compiler('c') }}
    - {{ compiler('fortran') }}
    - {{ compiler('cxx') }}
```

**Comment by beckermr** [comment]
OK. You likely need some of the patches here: https://github.com/conda-forge/cosmosis-standalone-feedstock/tree/master/recipe

I am looking at the logs and the code is not set up to compile how conda needs it to. It needs to look into PREFIX for linking and the compilers are in a different prefix.

**Comment by joezuntz** [comment]
It just uses $FC, $CC, etc, which seem to be set by conda.  The only thing that went wrong in this build seems to be the lapack not being there any more.  Is there a particular think you're looking at?

**Comment by beckermr** [comment]
Some errors.

```
Not building the minuit sampler as the you need the Minuit2 library
  (And to set the MINUIT2_LIB and MINUIT2_INC environment variables.)
  This is not generally a problem.
```

This one might be fine.

```
/tmp/pip-req-build-0y6b1fok/cosmosis
  make[1]: f77: Command not found
  make[1]: Entering directory `/tmp/pip-req-build-0y6b1fok/cosmosis/datablock'
  make[1]: f77: Command not found
```

This one looks serious.

For lapack, the linking path is wrong

```
echo LAPACKLIB is -L/home/conda/staged-recipes/build_artifacts/cosmosis_1624625346368/_build_env/lib -llapack
  LAPACKLIB is -L/home/conda/staged-recipes/build_artifacts/cosmosis_1624625346368/_build_env/lib -llapack

```

**Comment by beckermr** [comment]
I wonder if you need this build script

https://github.com/conda-forge/cosmosis-standalone-feedstock/blob/master/recipe/build.sh

**Comment by joezuntz** [comment]
```
/tmp/pip-req-build-0y6b1fok/cosmosis
  make[1]: f77: Command not found
  make[1]: Entering directory `/tmp/pip-req-build-0y6b1fok/cosmosis/datablock'
  make[1]: f77: Command not found
```
This is just it when it runs `python setup.py clean` - it fails to find a fortran compiler then, but that doesn't matter and it still goes on to clean.  It's not a problem, and happens after the actual build failiure, which is further up and to do with missing lapack.



```
Not building the minuit sampler as the you need the Minuit2 library
  (And to set the MINUIT2_LIB and MINUIT2_INC environment variables.)
  This is not generally a problem.
```
Yep, not a problem.  I could fix, but in general MINUIT2 seems to have disappeared from the CERN website so I'm wary of depending on it.


```
echo LAPACKLIB is -L/home/conda/staged-recipes/build_artifacts/cosmosis_1624625346368/_build_env/lib -llapack
  LAPACKLIB is -L/home/conda/staged-recipes/build_artifacts/cosmosis_1624625346368/_build_env/lib -llapack
```

isn't that because you removed lapack above?

**Comment by joezuntz** [comment]
The stuff from your build script that seemed to be needed I've ported to the main cosmosis setup.py

**Comment by beckermr** [comment]
> isn't that because you removed lapack above?

Yes it is, but the build was relying on incorrect behavior. At build time there are two conda envs around. The first one contains the compilers and other build tools. This is the one called `build`. The second env is called `host`. This env contains the things the build is supposed to link to and is where the software gets installed. Right now, the build is trying to link to things in build, when in reality it needs to link to the stuff in host. 

I checked your setup.py and you are using CONDA_PREFIX for linking which is wrong when running conda build. You need to be using PREFIX.

**Comment by beckermr** [comment]
You can detect this via something like

```
if os.environ.get("CONDA_BUILD", "0") == "1":
    prefix = "PREFIX"
else:
    prefix = "CONDA_PREFIX"
```

**Comment by joezuntz** [comment]
I see, thanks!

Will do.

**Comment by beckermr** [comment]
For minuit, you can use minuit2-standalone. This is in the root github account and is maintained. 

https://github.com/conda-forge/minuit2-feedstock

**Comment by joezuntz** [comment]
I get invalid ELF header on the shared lib that's built that links to minuit minuit. Wondering if I was compiling it wrong? Or did I have the minuit_standalone in the wrong place?

**Comment by beckermr** [comment]
The meta.yaml looks fine. I'd make sure you are using all of conda-forge's values for CFLAGS, LDFLAGS, CXXFLAGS etc.

**Comment by beckermr** [comment]
Glad to see the tests passing. I'd like to look through the build logs before merging. 

For the record, what we're the changes needed here?

**Comment by joezuntz** [comment]
Yep, thanks for your help!  Three issues - the first was the conda flags not being used in one place, as you said.  The second was openmp not being included properly, and the third was an error in the pip package that included a `.so` file by mistake.

File: recipes/cosmosis-build-standard-library/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+requirements:
+  host:
+    - python
+    - pip
+  run:
+    - python
+    - cosmosis
+    # For compiling standard librarymodules
+    - compilers
+    - {{ mpi }}
```
Comment (beckermr) [inline recipes/cosmosis-build-standard-library/meta.yaml:26]:
```suggestion
    - {{ mpi }}
    - llvm-openmp  # [osx]
    - libgomp  # [linux]
```

I think you may need these when building the standard library. I see `-fopenmp` flags without these libs.

**Review by beckermr — CHANGES_REQUESTED** [review]
One more thing and then I think we're good to go!

File: recipes/cosmosis-build-standard-library/meta.yaml (hunk 9 above)
Comment (beckermr) [reply recipes/cosmosis-build-standard-library/meta.yaml:26]: It appears these libs are getting pulled in anyways, but they are direct dependencies and so need to be listed explicitly.

**Review by beckermr — APPROVED** [review]
I added them. Once the build passes we can merge!

Thanks a bunch Joe!

**Comment by joezuntz** [comment]
Cool, thanks so much!

**Pull request merged**